hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java

```diff
@@ -1253,6 +1253,15 @@ default boolean balancer() throws IOException {
    */
   boolean balance() throws IOException;
 
+  /**
+   * Invoke the balancer in dry run mode. Will show plan but not actually move any regions.
+   * Can NOT run for various reasons. Check logs.
+   *
+   * @return <code>true</code> if dry run ran, <code>false</code> otherwise.
+   * @throws IOException if a remote or network exception occurs
+   */
+  boolean dryRunBalance() throws IOException;
```

Comment:
I'm not a big fan of adding a brand new API just for the `dry-run` mode. What about overloading the existing `balance()` method?

Reply:
I could do that. I didn't initially do that for a few reasons:

- There are already a lot of overloads of `balance()`, between the deprecated `balancer()` and the `boolean force`.
- I didn't want to add another boolean argument. IMO an interface `balance(boolean, boolean)` can be easily messed up. I could expose a client-side version of RunMode enum though, but in order to preserve backwards compatibility we'd have to add a whole set of new overloads.
- In the comments of HBASE-26147 we discussed enhancing dry run mode to accept config overrides for cost functions. This would make it much more intuitive to iterate on cost functions without doing a bunch of gymnastics. If we wanted to do that, we'd need to add another argument that would be specific to dry run mode which would further complicate the interface of `balance()`.

To the last reason, I could plan ahead for that and have `balance()` take a BalancerConfig object instead of an enum. That'd give us the future flexibility to add config overrides to just dry run mode, but it's a more complex interface and not _that_ much better than having 2 separate methods.

Do you have any opinions on the above? I'm not sure how the hbase devs way interface change questions like this.

Reply:
> I'm not sure how the hbase devs weigh interface change questions like this.

Yeah, that's exactly the problem ;). As you have added it now, you've added a new method to the public API which we would not be able to remove before HBase 4.0.0 at this point. Meaning, we want to be very cautious when adding new methods on public API.

I believe our specification technically allow us to define a single method on the interface to be at a lesser visibility (not Public), but it's not something we typically do, especially on one of the most user-facing interfaces.

Reply:
Ok, so it's more acceptable to modify the public API in this case? Here are some possibilities I can come up with:

- Modify the existing `balance(boolean force)` to add a new `boolean dryRun` argument. So `balance(boolean force, boolean dryRun)`. I don't love putting 2 booleans back to back like this, especially since force + boolean doesn't quite make sense.
  - In this case I think I'd actually have to add a new method to preserve backwards compatibility, and the the existing `balance(boolean force)` would delegate to `return balance(force, false)`.
- Change the going forward signature to a RunMode enum (or other name), with `REQUEST`, `FORCE`, `DRY_RUN`.
  - In this case I would need a new `balance(RunMode runMode)`, and then the existing balance would delegate to `return balance(force ? RunMode.FORCE : RunMode.REQUEST)`
- Change the going forward signature to a BalancerConfig object. We have a lot of options for how the BalancerConfig is constructed, either as a builder or factory methods.
  - Similar to above, I'd add a new `balance(BalancerConfig)`, and delegate based on the value of the force boolean.

I honestly think the last option is the best for long term as it sets us up for future expansion, but it's also the most complicated. I'd be fine to add that, but all of these options change the public API. I'm not sure which is the lesser of all evils from that perspective.

Reply:
Thinking about it more and reading the aspirational semver doc, it seems like the "lesser of all evils" is the change that a) doesn't break compatibility (all of them), and b) will have the most longevity. We don't want to add dryRunBalancer today if we plan to replace it in the "near future".

Given that, if you don't like `dryRunBalancer`, I would vote for BalancerConfig. We know that we want to add support for config overrides, and it'd be awkward to add that aside an enum or boolean.

I personally actually prefer `dryRunBalancer` because unlike `boolean force`, it feels like a distinct mode from the normal balancer and we might iterate on it over time separately from the normal balancer. For instance we might add return values that differ from what the normal balancer will ever return, arguments that don't make sense for the normal balancer, etc.

That said, I'm not a committer so my vote may not be worth that much :) I'd be open to doing which ever will get it merged.

Reply:
> I would vote for BalancerConfig. We know that we want to add support for config overrides, and it'd be awkward to add that aside an enum or boolean.

Yeah, I think this is the nicer of the options. This lets us avoid making big changes to `Admin` itself, but limit those to an `Evolving` `BalancerConfig` object. We provide a bit more stability to `Admin` that way, IMO.

Reply:
I am wondering if we will need to run balance() with modified configs. Will open a jira to follow up So I vote for a single API.

Reply:
https://issues.apache.org/jira/browse/HBASE-26177

Reply:
Thank you for the comment @clarax. I'm working on this new. I've decided to call it BalanceRequest, because that felt more natural (and also doesn't conflate configs that affect balancer like cost functions).

For now the API is very simple, so I decided against a builder. Instead it's a static factory for now, with 3 static methods:

- `#forExecution()` -- normal run
- `#forExecutionIgnoringRegionsInTransition()` -- normal run but ignore regions in transition. This is the replacement for `boolean force`. It's more verbose but also clearer what it does. IF we decide to add other force-like features in the future, we can give them each their own setter in a builder.
- `#forDryRun()` -- dry run mode

In HBASE-26177 we could add a new `forDryRun(CostConfigs)` or something. Or we can turn it into a builder at that point, if we decide that it makes sense to allow in both modes. For example, `BalancerRequest.forDryRun().withCostConfigs(CostConfigs).build()`

I'm just working on plumbing this all around now. Behind the scenes we are using the same RunMode enum I added in this PR, but that is an InterfaceAudience.PRIVATE, for internal use.